helper/schema/schema_test.go

```diff
@@ -3335,6 +3335,25 @@ func TestSchemaMap_Validate(t *testing.T) {
 				fmt.Errorf("\"optional_att\": conflicts with required_att (\"required-val\")"),
 			},
 		},
+
+		"Readonly attribute is set by user": {
+			Schema: map[string]*Schema{
+				"readonly_att": &Schema{
+					Type:     TypeString,
+					ReadOnly: true,
+					Required: true,
+				},
+			},
+
+			Config: map[string]interface{}{
+				"readonly_att": "user-set-val",
+			},
+
+			Err: true,
+			Errors: []error{
+				fmt.Errorf("\"readonly_att\": This field is readonly and cannot be set by users"),
+			},
+		},
```

Comment:
Can you add another test here "Readonly attribute with no config is OK", with `Config: nil, Err: false`?

That should force the code above to verify that the config is actually set (or prove that the code path isn't hit until we _have_ config already).